### PR TITLE
Use Deterministic Signatures (RFC 6979)

### PR DESCRIPTION
**Pull request by harryprotist**
Adds `signDeterministic` function to `sjcl.ecc.ecdsa.secretKey`, which uses a deterministically generated ephemeral key instead of the usual random one. 

Uses HMAC for generating the ephemeral key, as described in [RFC 6979](http://tools.ietf.org/html/rfc6979). It uses the implementation of HMAC from sjcl, and whatever hash is passed to the function (default sjcl.hash.sha256).

### CODE REVIEW

File: src/js/sjcl-custom/sjcl-ecdsa-signdeterministic.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/** Diffie-Hellmann function
+* @param {bitArray} hash hash to sign. 
+* @param {Object} type of hash used for hmac (default sjcl.hash.sha256)
+* @param {boolean} fakeLegacyVersion use old legacy version
+*/
+sjcl.ecc.ecdsa.secretKey.prototype.signDeterministic = function(hash, hash_object, fakeLegacyVersion) {
```
Comment (clark800) [inline src/js/sjcl-custom/sjcl-ecdsa-signdeterministic.js]: Remove fakeLegacyVersion

File: src/js/sjcl-custom/sjcl-ecdsa-signdeterministic.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+  var R = this._curve.r,
+      l = R.bitLength();
```
Comment (clark800) [inline src/js/sjcl-custom/sjcl-ecdsa-signdeterministic.js]: Remove `l` variable, move `R` before usage

File: src/js/sjcl-custom/sjcl-ecdsa-signdeterministic.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+  var hmac = new sjcl.misc.hmac(this.get(), hash_object || sjcl.hash.sha256);
```
Comment (clark800) [inline src/js/sjcl-custom/sjcl-ecdsa-signdeterministic.js]: var secretKey = this.get();

File: src/js/sjcl-custom/sjcl-ecdsa-signdeterministic.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  var k = hmac.encrypt(hash);
+  while (sjcl.bitArray.bitLength(k) < this._curveBitLength) {
```
Comment (clark800) [inline src/js/sjcl-custom/sjcl-ecdsa-signdeterministic.js]: Extract method `generateK`

File: src/js/sjcl-custom/sjcl-ecdsa-signdeterministic.js (hunk 1 above)
Comment (clark800) [inline src/js/sjcl-custom/sjcl-ecdsa-signdeterministic.js]: Call secretKey.sign to reduce duplication

Comment (clark800) [inline src/js/sjcl-custom/sjcl-ecdsa-signdeterministic.js]: Use camelcase

**Comment by sublimator** [comment]
@justmoon You probably wanna look at this

**Comment by harryprotist** [comment]
I applied all the suggestions except for "Call secretKey.sign to reduce duplication", because the `k_for_testing` parameter in secretKey.sign doesn't work

File: src/js/sjcl-custom/sjcl-ecdsa-signdeterministic.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+'use strict';
+/*global sjcl*/
+/*eslint-disable new-cap*/
+
+sjcl.ecc.ecdsa.secretKey.prototype.generateK =
```
Comment (clark800) [inline src/js/sjcl-custom/sjcl-ecdsa-signdeterministic.js]: I don't think this needs to be exposed publicly

**Pull request closed without merging**